execution_engine/src/storage/global_state/lmdb.rs

```diff
@@ -281,16 +289,58 @@ impl StateProvider for LmdbGlobalState {
         correlation_id: CorrelationId,
         trie_keys: Vec<Digest>,
     ) -> Result<Vec<Digest>, Self::Error> {
-        let txn = self.environment.create_read_txn()?;
-        let missing_descendants =
-            missing_trie_keys::<Key, StoredValue, lmdb::RoTransaction, LmdbTrieStore, Self::Error>(
+        if trie_keys
+            .iter()
+            .filter(|digest| {
+                !self
+                    .digests_without_missing_descendants
+                    .read()
+                    .unwrap()
+                    .contains(digest)
+            })
+            .count()
+            == 0
+        {
+            info!("no need to call missing_trie_keys");
+            Ok(vec![])
+        } else {
+            let txn = self.environment.create_read_txn()?;
+            let missing_descendants = missing_trie_keys::<
+                Key,
+                StoredValue,
+                lmdb::RoTransaction,
+                LmdbTrieStore,
+                Self::Error,
+            >(
                 correlation_id,
                 &txn,
                 self.trie_store.deref(),
-                trie_keys,
+                trie_keys.clone(),
+                &self.digests_without_missing_descendants.write().unwrap(),
             )?;
-        txn.commit()?;
-        Ok(missing_descendants)
+            if missing_descendants.is_empty() {
+                // There were no missing descendants on `trie_keys`, let's add them *and all of
+                // their descendents* to the cache.
+
+                let mut all_descendents: HashSet<Digest> =
```

Comment:
```suggestion
                let mut all_descendants: HashSet<Digest> =
```

Reply:
Fixed in aac1d6b4d9dba612378a1af8429fe70f09415908